### PR TITLE
Expose statistics on extrems to prometheus

### PR DESCRIPTION
**Pull request by hawkowl**
Fixes https://github.com/matrix-org/synapse/issues/5339

Exposes the rooms with forward extremities as a histogram to be scraped by Prometheus.

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -220,13 +221,35 @@ class EventsStore(
...
 
         self._event_persist_queue = _EventPeristenceQueue()
         self._state_resolution_handler = hs.get_state_resolution_handler()
 
+        # Collect metrics on the number of forward extremities that exist.
+        self._current_forward_extremities_amount = {}
+
+        BucketCollector(
+            "synapse_forward_extremities",
+            lambda: self._current_forward_extremities_amount,
```
Comment (erikjohnston) [inline synapse/storage/events.py]: You'll want to configure the buckets a bit here, something like `[1, 2, 3, 5, 7, 10, 15, 20, 50, 100, 200, 500]`, I think the default buckets are a lot larger.

File: synapse/storage/events.py [hunk 2]
```diff
@@ -220,13 +221,35 @@ class EventsStore(
...
+        # Read the extrems every 60 sec
+        hs.get_clock().looping_call(self._read_forward_extremities, 60000)
+
+    @defer.inlineCallbacks
+    def _read_forward_extremities(self):
+        def fetch(txn):
+            txn.execute(
+                "select room_id, count(*) c from event_forward_extremities "
+                "group by room_id"
+            )
```
Comment (erikjohnston) [inline synapse/storage/events.py]: (You can use mutltline string literals btw, which makes it much nicer to work with imo, as it allows easier c+p and you don't have to worry about remembering to add spaces)

---

This query takes ~25s on matrix.org from cold, so I'd vote we run it much less frequently (once an hour?).  I think if we're interested in more fine grained metrics we probably want to count the number of extremities of rooms that are active, e.g. count extremities for a room each time we do state res, count the number of extremities when we write events (we already pull out the extremities when we persist events), etc.

---

Let's also not bother pulling out the `room_id` if we're not going to use it.

File: synapse/storage/events.py (hunk 1 above)
Comment (hawkowl) [reply synapse/storage/events.py]: done

File: synapse/storage/events.py (hunk 2 above)
Comment (hawkowl) [reply synapse/storage/events.py]: done

File: synapse/metrics/__init__.py [hunk 3]
```diff
@@ -193,12 +191,58 @@ def _register_with_collector(self):
         all_gauges[self.name] = self
 
 
+@attr.s(hash=True)
+class BucketCollector(object):
+
```
Comment (erikjohnston) [inline synapse/metrics/__init__.py]: *cough* docstring *cough*

Comment (hawkowl) [reply synapse/metrics/__init__.py]: what r docstring?

(fixed)

**Pull request merged**